contracts/src/v0.8/l2ep/dev/SequencerUptimeFeed.sol

```diff
@@ -0,0 +1,234 @@
+
+/// @title ScrollSequencerUptimeFeed - L2 sequencer uptime status aggregator
+/// @notice L2 contract that receives status updates, and records a new answer if the status changed
+abstract contract SequencerUptimeFeed is AggregatorV2V3Interface, TypeAndVersionInterface, SimpleReadAccessController {
+  /// @dev Round info (for uptime history)
+  struct Round {
+    bool status;
+    uint64 startedAt;
+    uint64 updatedAt;
+  }
+
+  /// @dev Packed state struct to save sloads
+  struct FeedState {
+    uint80 latestRoundId;
+    bool latestStatus;
+    uint64 startedAt;
+    uint64 updatedAt;
+  }
+
+  /// @notice Sender is not the L2 messenger
+  error InvalidSender();
+  /// @notice Replacement for AggregatorV3Interface "No data present"
+  error NoDataPresent();
+  /// @notice Contract is not yet initialized
+  error Uninitialized();
+
+  /// @dev Emitted when the L1 sender is updated
+  event L1SenderTransferred(address indexed from, address indexed to);
+  /// @dev Emitted when an `updateStatus` call is ignored due to unchanged status or stale timestamp
+  event UpdateIgnored(bool latestStatus, uint64 latestTimestamp, bool incomingStatus, uint64 incomingTimestamp);
+  /// @dev Emitted when a updateStatus is called without the status changing
+  event RoundUpdated(int256 status, uint64 updatedAt);
+
+  // solhint-disable-next-line chainlink-solidity/all-caps-constant-storage-variables
+  uint8 public constant override decimals = 0;
+  // solhint-disable-next-line chainlink-solidity/all-caps-constant-storage-variables
+  string public constant override description = "L2 Sequencer Uptime Status Feed";
+  // solhint-disable-next-line chainlink-solidity/all-caps-constant-storage-variables
+  uint256 public constant override version = 1;
+
+  /// @dev L1 address
+  address private s_l1Sender;
+  /// @dev If true reverts if an invalid round was provided as an argument
+  bool private s_revertIfInvalidRound;
+
+  /// @dev s_latestRoundId == 0 means this contract is uninitialized.
+  FeedState internal s_feedState = FeedState({latestRoundId: 0, latestStatus: false, startedAt: 0, updatedAt: 0});
+
+  /// @dev mapping of round ID to round data
+  mapping(uint80 roundId => Round round) private s_rounds;
+
+  /// @param l1SenderAddress Address of the L1 contract that is permissioned to call this contract
+  /// @param revertIfInvalidRound if true, reverts if an invalid round ID is passed as input (otherwise returns 0 for some fields)
+  constructor(address l1SenderAddress, bool revertIfInvalidRound) {
+    _setL1Sender(l1SenderAddress);
+    s_revertIfInvalidRound = revertIfInvalidRound;
+  }
+
+  /// @notice Check that this contract is initialised, otherwise throw
+  modifier requireInitialized() virtual {
+    if (s_feedState.latestRoundId == 0) {
+      revert Uninitialized();
+    }
+    _;
+  }
+
+  /// @notice Should revert if the sender is not authorized to call `updateStatus`
+  modifier requireValidSender() virtual {
```

Comment:
I don't love virtual modifiers, they are very hard to reason about. Is this needed?

Reply:
It's not strictly needed - I've changed the implementation such that this is an internal virtual function as per your recommendation in Slack